### PR TITLE
demo of nixpkgs-fmt 35ae8e6498e23d1b05091976acc0ae2a6f749096

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

Get some feedback on the nixpkgs-fmt formatting engine. This is the output from the latest version of master.

###### Things done

**Not intended to be merged**

### CODE REVIEW

File: doc/shell.nix [hunk 1]
```diff
@@ -1,5 +1,6 @@
-{ pkgs ? import ../. {} }:
-(import ./default.nix {}).overrideAttrs (x: {
-  buildInputs = x.buildInputs ++ [ pkgs.xmloscopy pkgs.ruby ];
+{ pkgs ? import ../. { } }:
+(import ./default.nix { }).overrideAttrs
+  (x: {
```
Comment (zimbatm) [inline doc/shell.nix]: I think the original version was more readable.

File: flake.nix [hunk 2]
```diff
@@ -7,7 +7,6 @@
 
   outputs = { self }:
     let
-
```
Comment (zimbatm) [inline flake.nix]: as a general rule nixpkgs-fmt doesn't collapse lines, but not 100% sure in this case

File: flake.nix [hunk 3]
```diff
@@ -17,34 +16,36 @@
...
+                {
+                  system.nixos.versionSuffix =
                     ".${lib.substring 0 8 self.lastModified}.${self.shortRev or "dirty"}";
                   system.nixos.revision = lib.mkIf (self ? rev) self.rev;
                 }
               ];
-          });
-      };
+            }
+            );
```
Comment (zimbatm) [inline flake.nix]: these two should be collapsed. I think the rule is that if N containers are opened on the same line, then they should be closed on the same line as well

File: lib/attrsets.nix [hunk 4]
```diff
@@ -22,12 +20,14 @@ rec {
...
+      attr = head attrPath;
     in
-      if attrPath == [] then e
-      else if e ? ${attr}
-      then attrByPath (tail attrPath) default e.${attr}
-      else default;
+      if attrPath == [ ] then e
+      else
+        if e ? ${attr}
+        then attrByPath (tail attrPath) default e.${attr}
```
Comment (zimbatm) [inline lib/attrsets.nix]: not sure if it's worth the extra complexity to handle the `else if` case. the flat version looks nicer.

File: lib/attrsets.nix [hunk 5]
```diff
@@ -152,11 +158,13 @@ rec {
        => { a = [ 2 3 ]; }
   */
   foldAttrs = op: nul: list_of_attrs:
-    fold (n: a:
-        fold (name: o:
-          o // { ${name} = op n.${name} (a.${name} or nul); }
+    fold (
+      n: a:
```
Comment (zimbatm) [inline lib/attrsets.nix]: not sure why `n: a:` was pushed to the next line here. to be investigated.

File: lib/cli.nix [hunk 6]
```diff
@@ -42,42 +42,42 @@ rec {
...
-            else [ (mkOptionName k) (lib.generators.mkValueStringDefault {} v) ]
+  toGNUCommandLine =
+    {
+      # how to string-format the option name;
+      # by default one character is a short option (`-`),
+      # more than one characters a long option (`--`).
+      mkOptionName ? k:
+      if builtins.stringLength k == 1
+      then "-${k}"
+      else "--${k}"
```
Comment (zimbatm) [inline lib/cli.nix]: this should be indented

File: lib/cli.nix [hunk 7]
```diff
@@ -42,42 +42,42 @@ rec {
...
+      # though they can still appear as values of a list.
+      # By default, everything is printed verbatim and complex types
+      # are forbidden (lists, attrsets, functions). `null` values are omitted.
+      mkOption ? k: v:
+        if v == null
+        then [ ]
+        else [ (mkOptionName k) (lib.generators.mkValueStringDefault { } v) ]
     }:
-    options:
+      options:
```
Comment (zimbatm) [inline lib/cli.nix]: this should not be indented. function bodies stay on the same level.

File: lib/customisation.nix [hunk 8]
```diff
@@ -170,10 +174,11 @@ rec {
      garbage collection. */
   hydraJob = drv:
     let
-      outputs = drv.outputs or ["out"];
+      outputs = drv.outputs or [ "out" ];
 
       commonAttrs =
-        { inherit (drv) name system meta; inherit outputs; }
+        { inherit (drv) name system meta; inherit outputs;
+        }
```
Comment (zimbatm) [inline lib/customisation.nix]: not sure why `}` was pushed to the next line

**Comment by 7c6f434c** [comment]
Is evaluation failure caused by a merge problem, or by the original revision being broken?

Random observation: switch-like «else if» chains are better off without ever-increasing indent level. Not sure what is the best criterion; maybe «else» with the body containing only an «if» (possibly with another «else») should be indented as a special «else+if» thing?

File: lib/debug.nix [hunk 9]
```diff
@@ -95,17 +94,28 @@ rec {
...
-                        (const (modify (n - 1) fn)) v
-                      else v;
+    let
+      snip = v:
+        if isList v
+        then noQuotes "[…]" v
+        else
+          if isAttrs v
+          then noQuotes "{…}" v
+          else v;
```
Comment (zimbatm) [inline lib/debug.nix]: the original formatting was more readable to me

**Review by zimbatm — COMMENTED** [review]
a quick pass of things that look not so great

**Comment by zimbatm** [comment]
> Is evaluation failure caused by a merge problem, or by the original revision being broken?

this is a bug in the rnix parser that needs to be fixed. It looks like it's not parsing URLs exactly the same way yet and `https://git-ssb.celehner.com/%25n92DiQh7ietE%2BR%2BX%2FI403LQoyf2DtR3WQfCkDKlheQU%3D.sha256` is causing the problem.

File: lib/cli.nix (hunk 6 above)
Comment (7c6f434c) [reply lib/cli.nix]: You meant relative to `k:`, right?

**Comment by 7c6f434c** [comment]
> `https://git-ssb.celehner.com/%25n92DiQh7ietE%2BR%2BX%2FI403LQoyf2DtR3WQfCkDKlheQU%3D.sha256` is causing the problem

OK, so it is a good thing that the single-pass auto-quoting did not happen immediately…

File: lib/generators.nix [hunk 10]
```diff
@@ -144,90 +164,125 @@ rec {
...
+        else
+          if isString v
+          then ''"${libStr.escape [ ''"'' ] v}"''
+          else
+            if true == v
+            then "true"
+            else
+              if false == v
+              then "false"
+              else
```
Comment (7c6f434c) [inline lib/generators.nix]: This is where «else if» seems worth the complexity.

File: lib/attrsets.nix (hunk 4 above)
Comment (7c6f434c) [reply lib/attrsets.nix]: I agree this case would be better with `else if` but not worth it. I have commented on a larger example.

File: lib/generators.nix [hunk 11]
```diff
@@ -31,26 +30,44 @@ rec {
...
+                  then err "attrsets" v
+                    # functions can’t be printed of course
+                  else
+                    if isFunction v
+                    then err "functions" v
+                      # let’s not talk about floats. There is no sensible `toString` for them.
+                    else
+                      if isFloat v
+                      then err "floats" v
+                      else err "this value is" (toString v);
```
Comment (roberth) [inline lib/generators.nix]: Storage space for some cheese :cheese:

**Pull request closed without merging**